src/analysis/inner_product_space/positive.lean

```diff
@@ -32,21 +42,329 @@ of requiring self adjointness in the definition.
+
+/-- `U` is pairwise orthogonal to `Uᗮ` -/
+lemma submodule.orthogonal_is_pairwise_orthogonal (U : submodule 𝕜 E) :
+  U.pairwise_orthogonal Uᗮ :=
+λ u v, (set_like.coe_mem v) u (set_like.coe_mem u)
+
+/-- `U` is pairwise orthogonal to `V` if and only if `V` is pairwise
+orthogonal to `U` -/
+lemma submodule.pairwise_orthogonal_symm_iff (U V : submodule 𝕜 E) :
+  U.pairwise_orthogonal V ↔ V.pairwise_orthogonal U :=
+begin
+  simp_rw [submodule.pairwise_orthogonal, inner_eq_zero_sym],
+  rw forall_swap,
+end
+
+namespace linear_map
+
+/-- `T` is (semi-definite) **positive** if `T` is symmetric
+and `∀ x : V, 0 ≤ re ⟪x, T x⟫` -/
+def is_positive (T : E →ₗ[𝕜] E) : Prop :=
+T.is_symmetric ∧ ∀ x : E, 0 ≤ re ⟪x, T x⟫
+
+lemma is_positive_zero : (0 : E →ₗ[𝕜] E).is_positive :=
+begin
+  refine ⟨is_symmetric_zero, λ x, _⟩,
+  simp_rw [zero_apply, inner_re_zero_right],
+end
+
+lemma is_positive_one : (1 : E →ₗ[𝕜] E).is_positive :=
+⟨is_symmetric_id, λ x, inner_self_nonneg⟩
+
+lemma is_positive.add {S T : E →ₗ[𝕜] E} (hS : S.is_positive) (hT : T.is_positive) :
+  (S + T).is_positive :=
+begin
+  refine ⟨is_symmetric.add hS.1 hT.1, λ x, _⟩,
+  rw [add_apply, inner_add_right, map_add],
+  exact add_nonneg (hS.2 _) (hT.2 _),
+end
+
+lemma is_positive.inner_nonneg_left {T : E →ₗ[𝕜] E} (hT : is_positive T) (x : E) :
+  0 ≤ re ⟪T x, x⟫ :=
+by { rw inner_re_symm, exact hT.2 x, }
+
+lemma is_positive.inner_nonneg_right {T : E →ₗ[𝕜] E} (hT : is_positive T) (x : E) :
+  0 ≤ re ⟪x, T x⟫ :=
+hT.2 x
+
+/-- a linear projection onto `U` along its complement `V` is positive if
+and only if `U` and `V` are pairwise orthogonal -/
+lemma linear_proj_is_positive_iff {U V : submodule 𝕜 E} (hUV : is_compl U V) :
+  (U.subtype.comp (U.linear_proj_of_is_compl V hUV)).is_positive ↔ U.pairwise_orthogonal V :=
+begin
+  split,
+  { intros h u v,
+    rw [← submodule.linear_proj_of_is_compl_apply_left hUV u,
+        ← submodule.subtype_apply _, ← comp_apply, h.1 _ _,
+        comp_apply, submodule.linear_proj_of_is_compl_apply_right hUV v,
+        map_zero, inner_zero_right], },
+  { intro h,
+    have : (U.subtype.comp (U.linear_proj_of_is_compl V hUV)).is_symmetric,
+    { intros x y,
+      nth_rewrite 0 ← submodule.linear_proj_add_linear_proj_of_is_compl_eq_self hUV y,
+      nth_rewrite 1 ← submodule.linear_proj_add_linear_proj_of_is_compl_eq_self hUV x,
+      simp_rw [inner_add_right, inner_add_left, comp_apply,
+               submodule.subtype_apply _, h _ _, inner_eq_zero_sym.mp (h _ _)], },
+    refine ⟨this, _⟩,
+    intros x,
+    rw [comp_apply, submodule.subtype_apply,
+        ← submodule.linear_proj_of_is_compl_idempotent,
+        ← submodule.subtype_apply, ← comp_apply,
+        ← this _ ((U.linear_proj_of_is_compl V hUV) x)],
+    exact inner_self_nonneg, },
+end
+
+/-- set over `𝕜` is **non-negative** if all its elements are real and non-negative -/
+def set.is_nonneg (A : set 𝕜) : Prop :=
+∀ x : 𝕜, x ∈ A → ↑(re x) = x ∧ 0 ≤ re x
```

Comment:
This might be better written as `A ⊆ set.range (algebra_map ℝ≥0 𝕜)`

Reply:
gives me an error when I write `A ⊆ set.range (algebra_map ℝ≥0 𝕜)`

Reply:
What is the error?

Reply:
failed to synthesize type class instance for `has_le (Π (A : Type ?) [_inst_1 : comm_semiring ℝ] [_inst_2 : semiring A] [_inst_3 : algebra ℝ A], ℝ →+* A)`

Reply:
That error message is telling you that you forgot an argument somewhere. Can you show what you actually tried?

Reply:
I literally wrote what you wrote
```lean
def set.is_nonneg (A : set 𝕜) : Prop :=
A ⊆ set.range (algebra_map ℝ≥0 𝕜)
```

Reply:
Oh, you're missing `open_locale nnreal`